ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -64,6 +66,10 @@
   private final ForkChoiceNotifier forkChoiceNotifier;
   private final boolean proposerBoostEnabled;
 
+  private final Subscribers<OptimisticSyncSubscriber> optimisticSyncSubscribers =
+      Subscribers.create(true);
+  private final AtomicBoolean optimisticSyncing = new AtomicBoolean(false);
```

Comment:
This is only ever accessed from the fork choice thread so doesn't actually need to be thread safe. That said, the way the compareAndSet winds up working to only deliver events when the value changes is quite neat code so may be worth sticking with the AtomicBoolean for that simplicity...

Reply:
i was bouncing between boolean and AtomicBoolean, it isn't that evident that `onBlock` is called only from the fork choice thread... And TBH i'm still not convinced :)

Reply:
We're pretty screwed if it's called from any other thread. :)

The key is that it's called via:
```
.thenApplyAsync(
            payloadResult ->
                importBlockAndState(...),
            forkChoiceExecutor);
```

so it's the `forkChoiceExecutor` that will execute it and that ensures it's always on the event thread.

Reply:
ahh yes I missed that!